plug-ins/prometheus/architecture.md

```diff
@@ -0,0 +1,80 @@
+* `net_zlib_zlib` for [zlib](http://www.zlib.net/)
+
+The prometheus-cpp library will be brought in as a submodule of leap, but the Prometheus plugin will only use the core part of the library, specifically the part that implements the serialization. 
+
+## Plugin
+
+A new Prometheus plugin will be created.  This plugin will interface with other plugins (chain_plugin, producer_plugin, http_plugin) to supply the data to be exposed.  Instrumentation will be need to be added in the plugin to capture some of these statistics. 
+A metrics() call will be added to plugins to get their specific stats (in a non-prometheus dependent fashion).
+
+## Instrumentation of plugins
+
+Each of the various plugins will be instrumented to collect various runtime metrics.  Each of the plugins will have a a plugin_metrics struct that is accessed via a metrics() call which returns a reference to the individualized plugin_metrics. 
+The plugin_metrics will contain a set of member runtime_metrics which consist of a type {gauge or counter}, a name and a label.  In addition to the member metrics, a vector of metrics will be maintained for the use of the prometheus plugin. None of the non-prometheus plugins will have any dependency on the prometheus-cpp library. 
+Synchronization will be managed through the use of std::atomic.  This will ensure that all metric values are in a coherent state.  It does not ensure that all of the metric values represent a coherent 'snapshot' of a point (eg. it is possible that not all producer plugin values represent the state at a specific block).
+
+An initial phase of instrumentation will allow for the evaluation of the usability of the instrumentation when viewed from Prometheus.  The initial instrumentation will not include any per account or per contract metrics.  Additionally, the ability to expire metrics will not be included in the initial phase. To support dynamic metrics, the metric object for each plugin will include an indicator of when the set of metrics has last changed.  The following metrics will be collected for the initial phase of instrumentation
+
+## Net Plugin
+* gauge - number of clients
+* gauge - number of peers
+ 
+## Producer Plugin
+
+Producer plugin metrics will be gathered at the start of each block.
+
+* gauge - unapplied transaction queue sizes
+* counter - blacklisted transactions count
+* counter - blocks produced
+* counter - transactions produced
+* gauge - last irreversible
+* gauge - current block num
+* gauge - subjective billing account size
+* gauge - subjective billing block size
+* gauge - scheduled transactions
+ 
+## Chain Plugin
+* counter - snapshot rows processed
+* counter - replay block status 
+
+## Subsequent Instrumentation
+After the initial evaluation the following metrics will be added:
+
+* gauge - unapplied transaction queue
+  * byte size
+  * speculative CPU time
+* counter - blacklisted transactions
+  * by account
+  * by key count
+  * by contract count
```

Comment:
I don't think we want to report `by account` or `by key count` or `by contract count`. There could be a rather large list of each of those. We should report totals across all accounts and contracts.